File: backend/app/schemas/_budget_history.py
"""Budget history describes arrangements, never recalculated spending totals."""

from datetime import datetime
from typing import Literal

from pydantic import BaseModel, Field, field_serializer

from app.schemas._budgets import BudgetCategoryRequest
from app.schemas._money import NonNegativeMoneyMinor, SignedMoneyMinor
from app.services.time_service import to_iso


class BudgetSnapshot(BaseModel):
    home_currency_code: str | None
    total_amount_cents: NonNegativeMoneyMinor
    non_monthly_amount_cents: NonNegativeMoneyMinor
    rollover_amount_cents: SignedMoneyMinor
    excluded_categories: list[str]
    category_budgets: list[BudgetCategoryRequest]
    archived: bool


class BudgetRevisionResponse(BaseModel):
    row_version: int = Field(ge=1)
    change_kind: Literal["baseline", "create", "edit", "archive", "restore"]
    recorded_at: datetime
    snapshot: BudgetSnapshot

    @field_serializer("recorded_at")
    def serialize_time(self, value: datetime) -> str:
        return to_iso(value)


class BudgetHistoryResponse(BaseModel):
    ledger_id: str
    month: str
    items: list[BudgetRevisionResponse]
    next_before_version: int | None

File: android/app/src/main/java/com/ticketbox/domain/model/BudgetHistory.kt
package com.ticketbox.domain.model

data class BudgetHistoryPage(
    val ledgerId: String,
    val month: String,
    val items: List<BudgetRevision>,
    val nextBeforeVersion: Long?,
)

data class BudgetRevision(
    val rowVersion: Long,
    val changeKind: String,
    val recordedAt: String,
    val snapshot: BudgetArrangement,
)

data class BudgetArrangement(
    val homeCurrencyCode: String?,
    val totalAmountCents: Long,
    val nonMonthlyAmountCents: Long,
    val rolloverAmountCents: Long,
    val excludedCategories: List<String>,
    val categoryBudgets: List<BudgetCategoryArrangement>,
    val archived: Boolean,
)

data class BudgetCategoryArrangement(val category: String, val amountCents: Long)
